LMS Adjust Time Tracking Outliers

dandrabik opened the pull request:
## WHAT
We are doing nuanced time-tracking per prompt on the front end, but there is an existing bug in that calculation which is causing many time-tracking lists to have one element that is a very large number (likely from a resumed session). 

We should fix the bug on the frontend, but the code here is a backstop against clearly bad data being recorded. It replaces any individual time element that is 40x the median with the median, e.g. the total for the list `[1,2,3,4,999999]` would be calculated as `[1,2,3,4,3].sum`
## WHY
Time tracking is important to teachers, but these large outliers really distort the data. Replacing with the median of the other values seems like a reasonable stopgap.
## HOW
Add Array extension for `Array.median`. Add a bunch of tests. Replace individual outliers with median.

Edit:
I added a `TimeTrackingCleaner` class and am editing the granular time_tracking data, and I've returned the `timespent` calculation to just be a `sum`.

Note `40x median` was my initial guess at a reasonable outlier threshold. It's likely that could be lower.

### Screenshots
Examples (time is in seconds)
![Screen Shot 2022-04-06 at 12 30 24 PM](https://user-images.githubusercontent.com/1304933/162023107-23b2982f-ab2e-433b-996d-04c567644700.png)
![Screen Shot 2022-04-06 at 12 29 59 PM](https://user-images.githubusercontent.com/1304933/162023110-11a2206d-8fe5-4078-82f4-637a10f61a2e.png)
![Screen Shot 2022-04-06 at 12 30 12 PM](https://user-images.githubusercontent.com/1304933/162023111-8d7f74b9-5e71-4a8a-b35a-48ec569bf76c.png)


### Notion Card Links
https://www.notion.so/quill/ActivitySession-Time-Tracking-Outliers-Backend-mitigation-dbf94350104d460f8a6f8ce5c95799a4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |   'YES'.
Have you deployed to Staging? | (Possible answers:Not yet - deploying now!
Self-Review: Have you done an initial self-review of the code below on Github? | YEs
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

dandrabik committed:
Add Array.median function.

dandrabik committed:
Adjust time tracking to remove large outliers.

dandrabik committed:
Add comment.

dandrabik committed:
Lint fixes.

dandrabik committed:
Adjust method to work for params.

dandrabik committed:
Fix small nil error with error reporting.

dandrabik committed:
Combine lines.

dandrabik committed:
Remove unneeded nil check.

brendanshean reviewed (approved):
Nice work on this.  This seems like a reasonable stopgap.   If more robust outlier detection is desired, one approach would be to compute z-scores on the set of values and identifying `|z_score| > 3`.  (This means normalizing all the data and then finding values that are greater than 3 standard deviations from the mean).

emilia-friedberg reviewed (approved):
I really like this solution. Thank you for coming up with that.

dandrabik commented:
> Nice work on this. This seems like a reasonable stopgap. If more robust outlier detection is desired, one approach would be to compute z-scores on the set of values and identifying `|z_score| > 3`. (This means normalizing all the data and then finding values that are greater than 3 standard deviations from the mean).

I had initially dismissed some statistical methods like `z-score`/`t-score` since the sample size is so small (most are below 10 samples, some are as low as 5 samples) and the outliers are so different that it distorts the mean a lot, e.g. the screen shots below, but there may be ways to use them. 

I'm definitely interested in ways to make this more robust long-term. We do have a population of these answers, so we could construct tables as reference. Also `50x` the median is a huge outlier (it's a way to step our toe into this), but I could see us lowering that. I might ping you to pick your brain on some of these approaches.

![Screen Shot 2022-04-07 at 10 25 46 AM](https://user-images.githubusercontent.com/1304933/162225148-87f6e3b4-aad2-49a1-907a-8afe8acb9a8e.png)
![Screen Shot 2022-04-07 at 10 36 49 AM](https://user-images.githubusercontent.com/1304933/162225170-27177be7-aedf-44fb-a3d6-948a6969fd6e.png)

brendanshean commented:
In the example given the z-scores are:
![Screen Shot 2022-04-07 at 11 01 28 AM](https://user-images.githubusercontent.com/2057805/162230103-f316c236-9f7a-48ca-860b-09ed1f580573.png)

Which means you could make the band tighter like |z_score| > 2.
Or like you mention construct a reference table for mu and sigma when computing z-scores.

dandrabik committed:
Small refactors and add failing test.

dandrabik committed:
Add time_tracking_cleaner

dandrabik committed:
Refactors and cleanup.

dandrabik committed:
Fix typo.

dandrabik committed:
Use service pattern. Fix lint issues.

dandrabik committed:
Fix tests.

dandrabik committed:
Add comment, last minute cleanup.

dandrabik committed:
Lint

dandrabik commented:
So, I've changed this PR a bit, so re-requesting reviews. @emilia-friedberg @brendanshean these are the changes

1. The replacement/calculation logic is basically the same, `40x median`. I changed it from 50x to 40x after looking at some data. I'm open to using a different calculation like z-score, but I'd like to do a little more testing on that, if we go that route (it would be easy to swap in), and I think it's better to get this version out.
2. Instead of just editing the `activity_session.timespent` calculation, this edits the granular `activity_session.data['time_tracking']` data so that data will match the sum in `activity_session.timespent`. This feels more future proof, in case we use the granular data (we already use it internally). Note, I record any edits in `activity_session.data['time_tracking_edits']` in case we need to audit.

dandrabik committed:
Use ErrorNotifier.

brendanshean commented on services/QuillLMS/spec/models/activity_session_spec.rb:
```diff
@@ -864,9 +864,25 @@
+      expect(ActivitySession.calculate_timespent(nil, 'hello')).to be_nil
+    end
+
+    it 'should return timestamp of session passed in' do
```
should this be 'timespent'

dandrabik replied:
Ah, nice catch, will update.

brendanshean commented on services/QuillLMS/app/models/activity_session.rb:
```diff
@@ -158,8 +165,18 @@ def finished?
+  end
+
+  def self.calculate_timespent(activity_session, time_tracking)
+    timespent = activity_session&.timespent || time_tracking_sum(time_tracking)
```
I'm not sure I understand the disjunction here.  Assuming, the there is cleaned data, why would we prefer `activity_session&.timespent` over `time_tracking_sum(time_tracking)`

dandrabik replied:
So this is copying the existing code (and just moving it to the model). I agree, this might be better to change, but I think there may be some reason this was in place (time_tracking is new-ish, so I think it's to avoid overwriting older data). Anyway, I'm aiming to keep this changeset as small as possible, so will skip this potential refactor for now.

brendanshean commented on services/QuillLMS/app/lib/time_tracking_cleaner.rb:
```diff
@@ -0,0 +1,43 @@
+
+  # replace outlier time_tracking values with the list's median value
+  # if they are 40x larger than the median
+  # keep track of original data in data['time_tracking_edits']
```
Tracking edits is a great idea.

brendanshean reviewed (approved):
This feels like an improvement from the previous commit.  I have one question concerning the `calculate_timespent` method.  Otherwise, looks good to go.

I think using the median makes sense for now, but long term we should look into utilize population statistics for outlier detection. 

On a related note, even without using them for z-score calculations, I think it would be useful to compute landing(μ, σ), prompt_1(μ, σ), prompt_2(μ, σ), etc. as that might provide some interesting insight to individual items.

dandrabik committed:
Preserve edits on subsequent runs.

dandrabik committed:
Whitespace.

dandrabik committed:
Ensure median_value is an int, add tests.

dandrabik committed:
Lint.

dandrabik committed:
Fix typo.

dandrabik committed:
Ignore proofreader passage key in adjustment for now.

dandrabik committed:
Fix issue with permitting ‘time_tracking_edits’ key
